File: python/문제풀이/피보나치.py
def fibo(n):
    if n==1 or n==2:
        return 1
    return fibo(n-1)+fibo(n-2)

def fibo_topdown(n):
    if n==1 or n==2:
        return 1
    if d[n] != 0:
        return d[n]
    d[n] = fibo_topdown(n-1) + fibo_topdown(n-2)
    return d[n]

def fibo_bottomup(n):
    d[1] = 1
    d[2] = 1
    for i in range(3,n+1):
        d[i] = d[i-1] + d[i-2]
    return d[n]

n = int(input())

#print(fibo(n))

d = [0]*100
print(fibo_topdown(n))

d = [0]*100
print(fibo_bottomup(n))

File: JavaScript/다이나믹 프로그래밍/피보나치.js
const d = new Array(100);

// 탑다운 방식
const fibo_topdown = (x) => {
  if (x === 1 || x === 2) {
    return 1;
  }
  if (d[x]) {
    return d[x];
  }
  d[x] = fibo(x - 1) + fibo(x - 2);
  return d[x];
};

// 보텀업 방식
const fibo_bottomup = (n) => {
  d[1] = 1;
  d[2] = 1;
  for (let i = 3; i < n + 1; i++) {
    d[i] = d[i - 1] + d[i - 2];
  }
  return d[n];
};

const solution = () => {
  console.log(fibo_bottomup(4));
};
